### PR TITLE
[1095] Add callback to PCA Builder for parent activity/window

### PR DESCRIPTION
**Pull request by MarkZuber**
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1095

### CODE REVIEW

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs [hunk 1]
```diff
@@ -132,6 +132,21 @@ public PublicClientApplicationBuilder WithUseCorporateNetwork(bool useCorporateN
...
 #endif
 
+#if RUNTIME || NETSTANDARD_BUILDTIME
+        /// <summary>
+        ///  Sets a reference to the ViewController (if using Xamarin.iOS), Activity (if using Xamarin.Android)
+        ///  IWin32Window or IntPtr (if using .Net Framework). Used for invoking the browser.
+        /// </summary>
+        /// <remarks>Mandatory only on Android. Can also be set via the PublicClientApplcation builder.</remarks>
+        /// <param name="parentActivityOrWindowFunc">The parent as an object, so that it can be used from shared NetStandard assemblies</param>
+        /// <returns>The builder to chain the .With methods</returns>
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: >/// <returns>The builder to chain the .With methods</returns> [](start = 8, length = 61)

please add the specific android use case using the CurrentActivity (not sure that's the actual name ...)

File: tests/Microsoft.Identity.Test.Unit.net45/ApiConfigTests/AcquireTokenInteractiveBuilderTests.cs [hunk 2]
```diff
@@ -35,7 +35,7 @@ await _harness.SetupAsync()
         [TestMethod]
         public async Task TestAcquireTokenInteractiveBuilderAsync()
         {
-            await AcquireTokenInteractiveParameterBuilder.Create(_harness.Executor, MsalTestConstants.Scope)
+            await AcquireTokenInteractiveParameterBuilder.Create(_harness.Executor, null, MsalTestConstants.Scope)
```
Comment (henrik-me) [inline tests/Microsoft.Identity.Test.Unit.net45/ApiConfigTests/AcquireTokenInteractiveBuilderTests.cs]: >null,  [](start = 84, length = 6)

default to null? and have parameter last?

File: tests/Microsoft.Identity.Test.Unit.net45/ApiConfigTests/AcquireTokenInteractiveBuilderTests.cs [hunk 3]
```diff
@@ -147,14 +147,54 @@ public async Task TestAcquireTokenInteractive_EmbeddedAndSystemOptions_Async()
...
+                 AcquireTokenInteractiveParameterBuilder.Create(_harness.Executor, null, MsalTestConstants.Scope)
                                                          .WithSystemWebViewOptions(options)
                                                          .WithUseEmbeddedWebView(true)
                                                          .ExecuteAsync()
                                                         ).ConfigureAwait(false);
 
             Assert.AreEqual(MsalError.SystemWebviewOptionsNotApplicable, ex.ErrorCode);
         }
+
+        [TestMethod]
```
Comment (henrik-me) [inline tests/Microsoft.Identity.Test.Unit.net45/ApiConfigTests/AcquireTokenInteractiveBuilderTests.cs]: >[TestMethod] [](start = 8, length = 12)

Seems like some tests for the setting the activity call back on the PCA builder is missing?

**Comment by henrik-me** [comment]
e2e test failing

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs (hunk 1 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: Do we want to have an aka.ms link ? to Interactive with Xamarin?

File: tests/Microsoft.Identity.Test.Unit.net45/ApiConfigTests/AcquireTokenInteractiveBuilderTests.cs [hunk 4]
```diff
@@ -147,14 +147,54 @@ public async Task TestAcquireTokenInteractive_EmbeddedAndSystemOptions_Async()
...
+        public async Task TestAcquireTokenInteractive_ParentWindow_OnlyAtAcquireTokenBuilder_Async()
+        {
+            IntPtr parentWindowIntPtr = new IntPtr(12345);
+
+            await AcquireTokenInteractiveParameterBuilder.Create(_harness.Executor, null, MsalTestConstants.Scope)
+                                                         .WithParentActivityOrWindow(parentWindowIntPtr)
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit.net45/ApiConfigTests/AcquireTokenInteractiveBuilderTests.cs]: We can get some E2E test coverage on this by modifying an existing Android test

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenInteractiveParameterBuilder.cs [hunk 5]
```diff
@@ -66,6 +68,19 @@ internal AcquireTokenInteractiveParameterBuilder WithCurrentSynchronizationConte
             return this;
         }
 
+        internal AcquireTokenInteractiveParameterBuilder WithParentActivityOrWindowFunc(Func<object> parentActivityOrWindowFunc)
+        {
+#if RUNTIME || NETSTANDARD_BUILDTIME
+
+            if (parentActivityOrWindowFunc != null)
+            {
+                WithParentActivityOrWindow(parentActivityOrWindowFunc());
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Client/ApiConfig/AcquireTokenInteractiveParameterBuilder.cs]: What happens if both `WithParentActivityOrWindow` and `WithParentActivityOrWindowFunc` are used? We should also capture the desired behavior in an unit test

Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ApiConfig/AcquireTokenInteractiveParameterBuilder.cs]: There is a test for that.  See TestAcquireTokenInteractive_ParentWindow_WithCallbackFuncAndAcquireTokenBuilder_Async in this PR

**Comment by MarkZuber** [comment]
> e2e test failing

flaky.  follow up tests passed:  https://outlook-sdf.office.com/mail/inbox/id/AAQkAGUxOTk2MTk3LTU0YzQtNDQ2My1hYTZmLTFiNDNlZWZhNDBjYQAQADx5k6QdiI1tX2X2xncZiBI%3D

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs (hunk 1 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: @henrik-me  -- Can you please clarify what you mean about adding the specific android use case?  Do you want another explicit method for .WithActivity() only on Android?  Or are you asking for updates to the XML documentation for this method?

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs [hunk 6]
```diff
@@ -132,6 +132,21 @@ public PublicClientApplicationBuilder WithUseCorporateNetwork(bool useCorporateN
...
 
+#if RUNTIME || NETSTANDARD_BUILDTIME
+        /// <summary>
+        ///  Sets a reference to the ViewController (if using Xamarin.iOS), Activity (if using Xamarin.Android)
+        ///  IWin32Window or IntPtr (if using .Net Framework). Used for invoking the browser.
+        /// </summary>
+        /// <remarks>Mandatory only on Android. Can also be set via the AcquireTokenInteractive builder.</remarks>
+        /// <param name="parentActivityOrWindowFunc">The parent as an object, so that it can be used from shared NetStandard assemblies</param>
+        /// <returns>The builder to chain the .With methods</returns>
+        public PublicClientApplicationBuilder WithParentActivityOrWindow(Func<object> parentActivityOrWindowFunc)
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: maybe it's my lack of mastering the English language, but "Func" looks weird to me. Wouldn't want to use Callback in the public API ?

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs (hunk 1 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: @MarkZuber : I don't think we have a specific aka.ms link. Feel free to create one.

Comment (bgavrilMS) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: I believe what Henrik is trying to point at is https://github.com/jamesmontemagno/CurrentActivityPlugin

Comment (MarkZuber) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: So we want to document this guys' android plugin in our xml documentation for MSAL?

Comment (MarkZuber) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: Created https://aka.ms/msal-net-android-activity with extra docs around this.  Please review.  @jmprieur  @henrik-me @bgavrilMS

Comment (henrik-me) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: I'm asking for updates to the XML doc here. Perhaps that can fit into the wiki page. I recall Bogdan pointing out some of the reasoning behind this method and what people would generally want to do here. I thought it made sense to mention that in our docs. The closer to the developer the better, though some details may fit better in the wiki page.

---
In reply to: [302260651](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1261#discussion_r302260651) [](ancestors = 302260651)

**Comment by henrik-me** [comment]
Generally looks fine. Minor comment on ensuring we include the main scenario Bogdan was mentioning in the XML doc as well as in the wiki page

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs (hunk 1 above)
Comment (bgavrilMS) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: The guy works for MSFT, though I never quite understood why the plugin is not part of Xamarin, at least Xamarin.Extras

Comment (bgavrilMS) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: aka.ms LGTM!

Comment (jmprieur) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: Isn't it also mandatory on iOS when using a broker?

Comment (jennyf19) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: @jmprieur yes...it is mandatory on iOS when invoking broker.

**Pull request merged**